### PR TITLE
Optimization of mLastGPSInformation.satInfoComplete

### PR DESCRIPTION
**Pull request by bettellam**
satInfoComplete was set in GxGSV reading and became true each time the satellite reading of each constellation was completed. Therefore in the same NMEA cycle, it would become true 3 or 4 times. 
Now satInfoComplete is set to true only once at the end of the NMEA sequence (GxZDA)
@m-kuhn This will bring improvements to QField. I have found that the number of satellites in use / in view is not always correct. 
The GNSS data reading routine must also be synchronized with: mLastGPSInformation.satInfoComplete = true

### CODE REVIEW

File: src/core/gps/qgsnmeaconnection.cpp [hunk 1]
```diff
@@ -363,7 +371,7 @@ void QgsNmeaConnection::processGsvSentence( const char *data, int len )
     // clear() on GGA
 
     // for determining when to graph sat info
-    mLastGPSInformation.satInfoComplete = ( result.pack_index == result.pack_count );
+    // mLastGPSInformation.satInfoComplete = ( result.pack_index == result.pack_count );
```
Comment (nyalldawson) [inline src/core/gps/qgsnmeaconnection.cpp:374]:
```suggestion
```

File: src/core/gps/qgsnmeaconnection.cpp [hunk 2]
```diff
@@ -363,7 +371,7 @@ void QgsNmeaConnection::processGsvSentence( const char *data, int len )
...
```
Comment (nyalldawson) [inline src/core/gps/qgsnmeaconnection.cpp:375]: Remove dead code, don't just comment it out
```suggestion

```

**Comment by bettellam** [comment]
@nyalldawson , you can add this pull request to version 3.28

**Comment by nyalldawson** [comment]
@bettellam let's hold off merging this till https://github.com/qgis/QGIS/pull/50744 is merged, as that PR adds the framework for testing the NMEA classes. Then when that PR is merged you can update this one to add unit tests covering your changes. (This NMEA logic is complex, and well overdue for test coverage :upside_down_face: )

File: src/core/gps/qgsnmeaconnection.cpp [hunk 3]
```diff
@@ -363,8 +371,6 @@ void QgsNmeaConnection::processGsvSentence( const char *data, int len )
     // clear() on GGA
 
     // for determining when to graph sat info
-    mLastGPSInformation.satInfoComplete = ( result.pack_index == result.pack_count );
```
Comment (nyalldawson) [inline src/core/gps/qgsnmeaconnection.cpp:366]: What about for devices which don't send a ZDA sentence? (I just checked, and I don't receive these)

Comment (nyalldawson) [reply src/core/gps/qgsnmeaconnection.cpp:366]: Looking at this, ZDA is a time and date message, so I don't think using it to detect when the satellite info is complete is valid at all.

Perhaps the logic should be that as soon as a "GSV" sentence is received we consider the satellite info incomplete, until we receive the next non-GSV sentence?

Comment (bettellam) [reply src/core/gps/qgsnmeaconnection.cpp:366]: > What about for devices which don't send a ZDA sentence? (I just checked, and I don't receive these)

Generally it is from the configuration of the GNSS receiver that NMEA messages are enabled!

Comment (bettellam) [reply src/core/gps/qgsnmeaconnection.cpp:366]: > 

I had chosen ZDA for the simple reason that it is always the last in NMEA cycle!
Your GSV strategy is the simplest to implement.
There will be a 1 cycle delay in NMEA sequence.

**Review by bettellam — COMMENTED** [review]
@nyalldawson  what do you think?

File: src/core/gps/qgsnmeaconnection.cpp [hunk 4]
```diff
@@ -425,6 +424,14 @@ void QgsNmeaConnection::processVtgSentence( const char *data, int len )
 
 void QgsNmeaConnection::processGsaSentence( const char *data, int len )
 {
+  if ( mLastGPSInformation.satInfoComplete = true )
```
Comment (agiudiceandrea) [inline src/core/gps/qgsnmeaconnection.cpp:427]:
```suggestion
  if ( mLastGPSInformation.satInfoComplete == true )
```

File: src/core/gps/qgsnmeaconnection.cpp [hunk 5]
```diff
@@ -424,7 +424,7 @@ void QgsNmeaConnection::processVtgSentence( const char *data, int len )
 
 void QgsNmeaConnection::processGsaSentence( const char *data, int len )
 {
-  if ( mLastGPSInformation.satInfoComplete = true )
+  if ( mLastGPSInformation.satInfoComplete == true )
```
Comment (agiudiceandrea) [inline src/core/gps/qgsnmeaconnection.cpp:427]: `if ( mLastGPSInformation.satInfoComplete )` should be enough.

**Comment by bettellam** [comment]
Test OK
test_20221031_112700

**Comment by bettellam** [comment]
@nyalldawson Maybe there is an error in:
tests / src / core / testqgsnmeaconnection.cpp (113)
GxGSV must always give FALSE:
113 QVERIFY (! Info.satInfoComplete);

also GxGSA must always give FALSE:
143 QVERIFY (! Info.satInfoComplete);
173 QVERIFY (! Info.satInfoComplete);

I hope I have not written a stupid thing !!!

**Comment by nyalldawson** [comment]
@bettellam 

It's ok for you to update those tests with the new values -- they were added just to reflect the current state of the logic. Can you also add a test code were satInfoComplete is true too?

**Comment by bettellam** [comment]
@nyalldawson Ok, I messed around with test code, now they work!

**Comment by bettellam** [comment]
@nyalldawson The new GPS Tools interface and GPS Widget are very interesting.

I did some connection tests to my GNSS receiver.
I think there is a problem with Connect / Connecting / Disconnecting.

I am available to do connection tests for your new GPS tools updates.
Thank you

**Comment by nyalldawson** [comment]
> I think there is a problem with Connect / Connecting / Disconnecting.

Can you post steps to reproduce?

**Comment by bettellam** [comment]
> @bettellamquesto manca ancora un caso di prova in cui le informazioni sul satellite vengono completate

@nyalldawson void TestQgsNmeaConnection :: testBasic ()
does this already.
In my receiver NMEA sequence is:
GNRMC satInfoComplete = true
GNVTG satInfoComplete = true
GNGGA satInfoComplete = true 
GNGSA satInfoComplete = false (reset count)
GxGSV satInfoComplete = false
GNGST satInfoComplete = true
GNZDA satInfoComplete = true
if GNGST and GNZDA were before GNGSA the satellite count would be offset by one cycle.

**Comment by bettellam** [comment]
> > I think there is a problem with Connect / Connecting / Disconnecting.
> 
> Can you post steps to reproduce?

@nyalldawson Connect button:
Test_Connection_1

does not complete the procedure there are 2 error messages.
The receiver is connected and receives GNSS data on a regular basis,
but I can no longer disconnect.

If I turn off the receiver, it still remains in an error situation and never connects!
Test_Connection_2

**Comment by bettellam** [comment]
> > I think there is a problem with Connect / Connecting / Disconnecting.
> 
> Can you post steps to reproduce?
@nyalldawson No, it does not depend on Quality.

The Connect button is visible for a moment Connecting and the errors start !!
PS: the version is the QGis portable Win64 one created by Pull Request.
In the meantime I have restarted QGis 4-5 times, never with connection OK !!

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**